### PR TITLE
syscall: Test adjtimex() EFAULT testcase with raw syscall.

### PR DESCRIPTION
**Pull request by vinay110974**
adjtimex02.c:  Tested EFAULT cases with only raw sycall.
  	       EINVAL and EPERM are tested for both raw_syscall and libc_call.

Signed-off-by: vinay kumar <vinay.m.engg@gmail.com>

### CODE REVIEW

File: testcases/kernel/syscalls/adjtimex/adjtimex02.c [hunk 1]
```diff
@@ -10,115 +10,188 @@
...
+	.lowlimit = -512000L,
+	.delta = -1,
+	.exp_err = EINVAL,
+	},
+};
+
+static struct test_variants
+{
+	int (*adjtimex) (void *timex);
+	enum tst_timex_type type;
```
Comment (metan-ucw) [inline testcases/kernel/syscalls/adjtimex/adjtimex02.c:82]: This type is not used at all, no need to add it.

File: testcases/kernel/syscalls/adjtimex/adjtimex02.c [hunk 2]
```diff
@@ -10,115 +10,188 @@
...
+	char *desc;
+} variants[] = {
+#if (__NR_adjtimex != __LTP__NR_INVALID_SYSCALL)
+	{
+	.adjtimex = sys_adjtimex,.type = TST_KERN_OLD_TIMEX,.desc = "syscall with old kernel spec"},
+#endif
```
Comment (metan-ucw) [inline testcases/kernel/syscalls/adjtimex/adjtimex02.c:88]: We are missing the libc variant here.

File: testcases/kernel/syscalls/adjtimex/adjtimex02.c [hunk 3]
```diff
@@ -10,115 +10,188 @@
 #include <unistd.h>
 #include <pwd.h>
 #include "tst_test.h"
+#include "lapi/syscalls.h"
 
 #define SET_MODE ( ADJ_OFFSET | ADJ_FREQUENCY | ADJ_MAXERROR | ADJ_ESTERROR | \
-	ADJ_STATUS | ADJ_TIMECONST | ADJ_TICK )
+				ADJ_STATUS | ADJ_TIMECONST | ADJ_TICK )
```
Comment (metan-ucw) [inline testcases/kernel/syscalls/adjtimex/adjtimex02.c:16]: This is a useless whitespace change, as a matter of fact the patch is polluted with changes like these and in many cases newly added code has completely wrong indentation. You can use checkpatch.pl shipped with Linux kernel source to check your code for such mistakes.

File: testcases/kernel/syscalls/adjtimex/adjtimex02.c [hunk 4]
```diff
@@ -10,115 +10,188 @@
...
+	}
+
+	if(tc[i].exp_err == EFAULT ) {
+			tst_res(TINFO, "EFAULT is tested only for raw syscall");
+			bufp = (struct timex *) -1;
+			TEST (tv->adjtimex (bufp));
+		} else {
+			TEST(tv->adjtimex(bufp));
+			TEST(adjtimex(bufp));
 	}
```
Comment (metan-ucw) [inline testcases/kernel/syscalls/adjtimex/adjtimex02.c:137]: Looks like moving the test data into the struct testcase structure made the code much worse. Why don't we just add the variants and keep the code mostly as it was? I guess that all that will be needed would be calling the tv->adjtimex() instead.

**Review by metan-ucw — CHANGES_REQUESTED** [review]

File: testcases/kernel/syscalls/adjtimex/adjtimex02.c (hunk 2 above)
Comment (vinay110974) [reply testcases/kernel/syscalls/adjtimex/adjtimex02.c:87]: Added libc variant for adjtimex.

File: testcases/kernel/syscalls/adjtimex/adjtimex02.c (hunk 4 above)
Comment (vinay110974) [reply testcases/kernel/syscalls/adjtimex/adjtimex02.c:132]: Corrections made.

**Comment by vinay110974** [comment]
Output of checkpatch.pl with adjtimex02.c after fixing indentation errors and warnings :

total: 0 errors, 0 warnings, 190 lines checked
ltp/testcases/kernel/syscalls/adjtimex/adjtimex02.c has no obvious style problems and is ready for submission.

File: testcases/kernel/syscalls/adjtimex/adjtimex02.c (hunk 3 above)
Comment (vinay110974) [reply testcases/kernel/syscalls/adjtimex/adjtimex02.c]: Verified with checkpatch.pl

File: testcases/kernel/syscalls/adjtimex/adjtimex02.c (hunk 1 above)
Comment (vinay110974) [reply testcases/kernel/syscalls/adjtimex/adjtimex02.c]: Removed.

File: testcases/kernel/syscalls/adjtimex/adjtimex02.c [hunk 5]
```diff
@@ -10,115 +10,185 @@
...
+static struct test_case{
+	unsigned int modes;
+	long lowlimit;
+	long highlimit;
+	long delta;
+	int exp_err;
+	int droproot;
+};
+
+static inline int libc_adjtimex(void *timex)
```
Comment (metan-ucw) [inline testcases/kernel/syscalls/adjtimex/adjtimex02.c:32]: You are storing a pointer to this function into a structure the inline keyword is complete nonsense here.

File: testcases/kernel/syscalls/adjtimex/adjtimex02.c [hunk 6]
```diff
@@ -10,115 +10,185 @@
...
-
-static struct timex *tim_save;
-static struct timex *buf;
+static int hz;		/* HZ from sysconf */
 
+static struct timex *tim_save, *buf;
 static struct passwd *ltpuser;
 
-static void verify_adjtimex(unsigned int nr)
+static struct test_case{
```
Comment (metan-ucw) [inline testcases/kernel/syscalls/adjtimex/adjtimex02.c:23]: The static does not make sense if the definition is not followed by a declaration.

Also there should be space before {

File: testcases/kernel/syscalls/adjtimex/adjtimex02.c [hunk 7]
```diff
@@ -10,115 +10,185 @@
...
+			tst_res(TCONF, "EFAULT is skipped for libc variant");
+			return;
+		}
+	}
+
+	TEST(tv->adjtimex(bufp));
+
+	if (tc[i].exp_err != TST_ERR) {
+		tst_res(TFAIL | TTERRNO, "adjtimex(): expected %d but failed with %d mode %x"
+				, tc[i].exp_err, TST_ERR, tc[i].modes);
```
Comment (metan-ucw) [inline testcases/kernel/syscalls/adjtimex/adjtimex02.c:138]: The coma should be at the end of the string here.

Also TTERRNO flag prints TST_ERR already so there is no point in printing it for a second time.

And you should use tst_strerrno() for the exp_err here as well.

File: testcases/kernel/syscalls/adjtimex/adjtimex02.c [hunk 8]
```diff
@@ -10,115 +10,185 @@
...
+		if (tc[i].modes == ADJ_TICK) {
+			tc[i].highlimit /= hz;
+			tc[i].lowlimit /= hz;
+		}
+	}
+
 	if ((adjtimex(tim_save)) == -1)
 		tst_brk(TBROK | TERRNO,
-			"adjtimex(): failed to save current params");
+		"adjtimex(): failed to save current params");
```
Comment (metan-ucw) [inline testcases/kernel/syscalls/adjtimex/adjtimex02.c:169]: LKML coding style prefers curly braces around multiline blocks so if the function call that is inside if has more than one line it should be enclosed in {}.

File: testcases/kernel/syscalls/adjtimex/adjtimex02.c [hunk 9]
```diff
@@ -10,115 +10,185 @@
...
 static void cleanup(void)
 {
+
 	tim_save->modes = SET_MODE;
 
-	/* Restore saved parameters */
 	if ((adjtimex(tim_save)) == -1)
-		tst_res(TWARN, "Failed to restore saved parameters");
+		tst_brk(TBROK | TERRNO,
+		"adjtimex(): failed to save current params");
```
Comment (metan-ucw) [inline testcases/kernel/syscalls/adjtimex/adjtimex02.c:179]: Here as well.

File: testcases/kernel/syscalls/adjtimex/adjtimex02.c [hunk 10]
```diff
@@ -10,115 +10,185 @@
...
 
-	/* clean up after ourselves */
-	if (nr == 3)
+	if (tc[i].exp_err == EFAULT) {
+		if (tv->adjtimex != libc_adjtimex)
+			bufp = (struct timex *) -1;
+		else {
+			tst_res(TCONF, "EFAULT is skipped for libc variant");
+			return;
+		}
```
Comment (metan-ucw) [inline testcases/kernel/syscalls/adjtimex/adjtimex02.c:131]: LKML coding style prefers curly braces around both if else blocks if they need to be around one of them.

File: testcases/kernel/syscalls/adjtimex/adjtimex02.c [hunk 11]
```diff
@@ -10,115 +10,185 @@
...
-		break;
-	case 2:
-		buf->tick = 1100000 / hz + 1;
-		expected_errno = EINVAL;
-		break;
-	case 3:
+
+	if (tc[i].droproot) {
 		/* Switch to nobody user for correct error code collection */
 		ltpuser = SAFE_GETPWNAM("nobody");
```
Comment (metan-ucw) [inline testcases/kernel/syscalls/adjtimex/adjtimex02.c:101]: This GETPWNAM() should be done once in the test setup instead in each test iteration.

File: testcases/kernel/syscalls/adjtimex/adjtimex02.c [hunk 12]
```diff
@@ -10,115 +10,185 @@
...
+
+static inline int libc_adjtimex(void *timex)
+{
+	return adjtimex(timex);
+}
+
+static inline int sys_adjtimex(void *timex)
+{
+	return tst_syscall(__NR_adjtimex, timex);
+}
```
Comment (metan-ucw) [inline testcases/kernel/syscalls/adjtimex/adjtimex02.c:40]: Here as well.

File: testcases/kernel/syscalls/adjtimex/adjtimex02.c [hunk 13]
```diff
@@ -10,115 +10,185 @@
...
+
+static struct test_variants
+{
+	int (*adjtimex)(void *timex);
+	char *desc;
+} variants[] = {
+	{ .adjtimex = libc_adjtimex, .desc = "vDSO or syscall with libc spec"},
+
+#if (__NR_adjtimex != __LTP__NR_INVALID_SYSCALL)
+	{ .adjtimex = sys_adjtimex,  .desc = "syscall with old kernel spec"},
```
Comment (metan-ucw) [inline testcases/kernel/syscalls/adjtimex/adjtimex02.c:86]: As far as I can tell this is not "old kernel spec".

This description should just say "__NR_adjtimex syscall" and the libc desc should say something as "libc adjtimex()"

**Review by metan-ucw — CHANGES_REQUESTED** [review]
Generally looks better but there are a few more things to fix.

**Comment by vinay110974** [comment]
Hi @metan-ucw,

Thanks for the review. The output of adjtimex02.c on qemuarm with corrections  (https://github.com/linux-test-project/ltp/commit/fdccc959e9736aee9c7ee055c1d3eb401b5e2cf2 ) is as below,

	root@qemuarm:~# ./adjtimex02 
	tst_buffers.c:55: TINFO: Test is using guarded buffers
	tst_test.c:1311: TINFO: Timeout per run is 0h 05m 00s
	adjtimex02.c:140: TINFO: Testing variant: libc adjtimex()
	adjtimex02.c:154: TPASS: adjtimex() error EPERM 
	adjtimex02.c:117: TCONF: EFAULT is skipped for libc variant
	adjtimex02.c:130: TPASS: adjtimex() error EINVAL
	adjtimex02.c:130: TPASS: adjtimex() error EINVAL
	adjtimex02.c:107: TCONF: this kernel normalizes buf. offset value if it is outside the acceptable range.
	adjtimex02.c:103: TCONF: this kernel normalizes buf. offset value if it is outside the acceptable range.
	tst_test.c:1311: TINFO: Timeout per run is 0h 05m 00s
	adjtimex02.c:140: TINFO: Testing variant: __NR_adjtimex syscall
	adjtimex02.c:154: TPASS: adjtimex() error EPERM 
	adjtimex02.c:130: TPASS: adjtimex() error EFAULT
	adjtimex02.c:130: TPASS: adjtimex() error EINVAL
	adjtimex02.c:130: TPASS: adjtimex() error EINVAL
	adjtimex02.c:107: TCONF: this kernel normalizes buf. offset value if it is outside the acceptable range.
	adjtimex02.c:103: TCONF: this kernel normalizes buf. offset value if it is outside the acceptable range.

File: testcases/kernel/syscalls/adjtimex/adjtimex02.c (hunk 11 above)
Comment (vinay110974) [reply testcases/kernel/syscalls/adjtimex/adjtimex02.c]: Below method is from the original adjtimex02.c. So removed "droproot" from "test_case" structure.

        buf->modes = SET_MODE;

        /* Switch to nobody user for correct error code collection */
        ltpuser = SAFE_GETPWNAM("nobody");
        SAFE_SETEUID(ltpuser->pw_uid);
        expected_errno = EPERM;

        TEST(tv->adjtimex(buf));

        if ((TST_RET == -1) && (TST_ERR == expected_errno)) {
                tst_res(TPASS, "adjtimex() error %s ",
                                tst_strerrno(expected_errno));
        } else {
                tst_res(TFAIL | TTERRNO,
                                "adjtimex(): returned %ld", TST_RET);
        }

        SAFE_SETEUID(0);

File: testcases/kernel/syscalls/adjtimex/adjtimex02.c (hunk 6 above)
Comment (vinay110974) [reply testcases/kernel/syscalls/adjtimex/adjtimex02.c]: Removed "static" from "test_case" structure and added a space before curly braces.

File: testcases/kernel/syscalls/adjtimex/adjtimex02.c (hunk 7 above)
Comment (vinay110974) [reply testcases/kernel/syscalls/adjtimex/adjtimex02.c]: Corrections made.

        if (tc[i].exp_err != TST_ERR) {
                tst_res(TFAIL | TTERRNO, "adjtimex(): expected %s mode %x",
                                        tst_strerrno(tc[i].exp_err), tc[i].modes);
                return;
        }

File: testcases/kernel/syscalls/adjtimex/adjtimex02.c (hunk 8 above)
Comment (vinay110974) [reply testcases/kernel/syscalls/adjtimex/adjtimex02.c:175]: Added curly braces for the if block.

File: testcases/kernel/syscalls/adjtimex/adjtimex02.c (hunk 9 above)
Comment (vinay110974) [reply testcases/kernel/syscalls/adjtimex/adjtimex02.c:186]: Added curly braces for the if block.

File: testcases/kernel/syscalls/adjtimex/adjtimex02.c (hunk 10 above)
Comment (vinay110974) [reply testcases/kernel/syscalls/adjtimex/adjtimex02.c:119]: Added curly braces for the "if" block also.

         if (tc[i].exp_err == EFAULT) {
                if (tv->adjtimex != libc_adjtimex) {
                        bufp = (struct timex *) -1;
                } else {
                        tst_res(TCONF, "EFAULT is skipped for libc variant");
                        return;
                }
        }

**Comment by metan-ucw** [comment]
I see that you have send a patch to the mailing list, can we close this and continue there?

**Pull request closed without merging**